README
= README for cvs-fast-export =

This program analyzes a collection of RCS files in a CVS repository
(or outside of one) and, when possible, emits an equivalent history in
the form of a fast-import stream.  Not all possible histories can be
rendered this way; the program tries to emit useful warnings when it
can't.  The program can also produce a visualization of the resulting
commit DAG in the DOT format handled by the graphviz suite.

This program could have been called rcs-fast-export with equal
appropriateness; the cvs-fast-export name was chosen to avoid
colliding with a pre-existing script with that name by Giuseppe
Bilotta.

The analysis stage of this code originally travelled as "parsecvs" 
and was written by Keith Packard in early 2006.  It was briefly 
maintained by Bart Massey before passing to Eric S. Raymond in 
late 2012; ESR wrote the fast-export output stage and renamed the
program to reflect its new function. More historical details are
in hacking.asc.

The distribution includes a tool, cvssync, for fetching masters from
CVS remote repositories so cvs-fast-export can see them.  You will
need rsync installed to use it.

A wrapper script called cvsconvert runs a conversion to git and
looks for content mismatches with the original CVS.

Also included is a tool called cvsreduce that strips content out of
trees of RCS/CVS masters, leaving only metadata structure in place. If
you encounter a bug in this program, sending the maintainer a reduced
version of your CVS tree greatly decreases the expected time to fix.

A more detailed guide to effective bug reporting is at reporting-bugs.asc.

This code has a regression-test suite; invoke it with "make check".
You will need RCS and CVS installed to run the tests, but they're
not required for production use of cvs-fast-export.  Installed CVS
is required to use the cvsconvert wrapper script.

See also the NEWS and TODO files.  There is an (incomplete) tour of
the internals in hacking.asc.


NEWS
		cvs-fast-export project news

Repository head:
    Tagged branchlets are created for any CVS tag not matching a gitspace commit.
    Many portability fixes for *BSD.

1.29: 2014-12-17
    A significant improvement in the correctness of vendor-branch handling.
    More speed improvements.

1.28: 2014-12-08
    Fix for buggy emission of lines ending in @.
    More speedups, and dramatic reduction in working-set size.

1.27: 2014-11-27
    The bug that produced "unnumbered head" warnings has been fixed.

1.26: 2014-11-05
    cvscompare changed to cvsconvert, a validating wrapper script.
    The rules for keyword expansion have changed yet again.

1.25: 2014-11-03
    Simplify and fully document -k; it now requires an argument.
    In normal (non-promiscuous) mode, paths containing CVSROOT are ignored.
    There is a new wrapper script, 'cvscompare', for sanity-checking conversions.
    A HOWTO on reporting bugs has been added to the distribution.

1.24: 2014-10-30
    Bugfix release: a late change to I/O buffering in 1.23 was faulty.

1.23: 2014-10-29
    New -l option for redirecting logs during long conversions.
    More speedups and working-set reductions.
    Incremental dumping can now be done in fast mode.

1.22: 2014-10-21
    New --embed-id and --expand options by Robert deBath.

1.21: 2014-10-19
    When incremental dumping, suppress tags associated with old commits.
    Performance improved by x8; see also the new -F and -C options.
    New -a optio to dump a list of author IDs found in a repo.

1.20: 2014-10-08
    Files not ending with ,v are now ignored unless the new -P option is on.
    New -t option for parallelizing analysis on multicore systems.

1.19: 2014-10-04
    Dramatic speedups in some bottleneck functions.

1.18: 2014-10-02
    A bug in the regression tests was fixed by slowing down calls to CVS.

1.17: 2014-09-11
    Some changes to reduce working-set size. More internals documentation.

1.16: 2014-09-04
    Added an internals tour to the documentation. Polished some comments.
    Prevented a possible buffer overrun.  Fixed broken -R option.

1.15: 2014-09-02
    Fixed a nasty order-instability bug that was confounding testing.
    Add a fatal error check for when revision numbers in input get too long.
    A significant speed improvement by tuning one of the sort algorithms.

1.14: 2014-08-12
    Fixed several issues near ignore conversions.

1.13: 2014-08-11
    Allow ()<> in symbol names.
    Fix a minor memory leak.
    Make cvssync a bit more liberal about SourceForge hostnames.
    In cvssync, leading /cvsroot can be omitted on Sourceforge paths.
    In cvssync, leading /sources can be omitted on Savannah paths.

1.12: 2014-06-26
    Allow []! in symbol names.
    Python in the test suite now runs under 2.6.

1.11: 2014-06-06
    Teach cvssync about sourceware.org. 

1.10: 2014-03-28
    Prepends CVS default ignores to converted .cvsignores.
    Generates a .gitignore containing default CVS ignores when necessary.

1.9: 2014-03-08
    RCS/CVS usernames may now begin with a digit (patch by Jesse Weinstein).
    Cleaned up a mess around keyword expansion; it had been misdocumented.
    Fixed some minor bugs in the test suite.
    Fixed a signedness issue on 64-bit machines.

1.8: 2014-02-19
    Processing of hardlinks field was incorrect, is now fixed. 

1.7: 2014-02-18
    Allow # in symbol names.  Addresses a corner case in the NetBSD repository.

1.6: 2014-02-17
    Fix a brown-paper-bag bug in cvssync.

1.5: 2014-02-16
    Skip a hardlinks field, if present.
    Add a -c option to cvssync that makes mirrors with CVSROOTs.

1.4: 2014-02-04
    Optimization to call sbrk() less often needs to be conditioned on GCC.

1.3: 2014-01-23
    Improved performance on masters with lots of symbols; thank Jens Bethkowsky.

1.2: 2014-01-04
    Smarter blob directory creation to reduce search overhead for blobs.

1.1: 2014-01-03
    Cope with CVS-NT kopt strings containing garbage binary data.

1.0: 2013-12-28
    Bug fix: Non-top-level .cvsignores are now converted.
    More speed and significant lowering of memory usage.
    Most error messages are now explained on the manual page.

0.8: 2013-12-16
    Processing speed has approximately tripled since last release.
    cvs-fast-export has save progress metering again.
    cvssync can now take a CVS URL argument.

0.7: 2013-12-15
    -i option for incremental dumping.
    -p option for enabling load status reports.
    Many documentation improvements.
    There is now a regression-test suite for the package.
    cvssync tool for mirroring remote CVS repos added.

0.6: 2013-12-10
    Improvements for CVS-NT support.
    Improve timing of missing-commitids message

0.5: 2013-05-21
    CVS-NT support. Code is Coverity-clean.

0.4: 2013-05-16
    Fix buggy handling of -k option.
    Add --remote (-e) and --stripprefix (-s) options.
    Avoid recursion overflow on very large repos.
    The fast-import stream ops are now emitted in the same order git uses.

0.3: 2013-01-16
    Fix a bonehead packaging error.

0.2: 2012-01-12
    Code revamped to emit a fast-export stream. Manual page added.

0.1: 2006-03-09
    Original code by Keith Packard; traveled as 'parsecvs'.


main.c
/*
 *  Copyright © 2006 Keith Packard <keithp@keithp.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or (at
 *  your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */
#include "cvs.h"
#include <unistd.h>
#include <getopt.h>
#include <regex.h>
#include <time.h>
#include <sys/stat.h>
#include <sys/resource.h>
#include "revdir.h"
#if defined(__GLIBC__)
#include <malloc.h>
#endif /* __GLIBC__ */

/* options */
int commit_time_window = 300;
bool progress = false;
FILE *LOGFILE;
#ifdef THREADS
int threads = NO_MAX;
#endif /* THREADS */

static import_options_t import_options = {
    .striplen = -1,
};

static int get_int_substr(const char * str, const regmatch_t * p)
{
    char buff[256];
    if (p->rm_so == -1)
	    return 0;
    if (p->rm_eo - p->rm_so >= sizeof(buff))
	    return 0;
    memcpy(buff, str + p->rm_so, p->rm_eo - p->rm_so);
    buff[p->rm_eo - p->rm_so] = 0;
    return atoi(buff);
}

static time_t mktime_utc(const struct tm * tm, const char* tzbuf)
{
    /* coverity[tainted_string_return_content] */
    char * old_tz = getenv("TZ");
    time_t ret;

    setenv("TZ", tzbuf, 1);

    tzset();

    ret = mktime((struct tm *)tm);

    if (old_tz)
	setenv("TZ", old_tz, 1);
    else
	unsetenv("TZ");

    tzset();

    return ret;
}

static time_t convert_date(const char *dte)
/* accept a date in anything close to RFC3339 form */
{
    static regex_t date_re;
    static bool init_re;

#define MAX_MATCH 16
    size_t nmatch = MAX_MATCH;
    regmatch_t match[MAX_MATCH];

    if (!init_re)
    {
	if (regcomp(&date_re, "([0-9]{4})[-/]([0-9]{2})[-/]([0-9]{2})[ T]([0-9]{2}):([0-9]{2}):([0-9]{2})( [-+][0-9]{4})?", REG_EXTENDED)) 
	    fatal_error("date regex compilation error\n");
	init_re = true;
    }

    if (regexec(&date_re, dte, nmatch, match, 0) == 0)
    {
	regmatch_t * pm = match;
	struct tm tm = {0};
	char tzbuf[32];
	int offseth, offsetm;

	/* first regmatch_t is match location of entire re */
	pm++;

	tm.tm_year = get_int_substr(dte, pm++);
	tm.tm_mon  = get_int_substr(dte, pm++);
	tm.tm_mday = get_int_substr(dte, pm++);
	tm.tm_hour = get_int_substr(dte, pm++);
	tm.tm_min  = get_int_substr(dte, pm++);
	tm.tm_sec  = get_int_substr(dte, pm++);
	offseth    = -get_int_substr(dte, pm++);

	offsetm = offseth % 100;
	if (offsetm < 0)
	    offsetm *= -1;
	offseth /= 100;
	snprintf(tzbuf, sizeof(tzbuf), "UTC%+d:%d", offseth, offsetm);

	tm.tm_year -= 1900;
	tm.tm_mon--;

	return mktime_utc(&tm, tzbuf);
    }
    else
    {
	return atoi(dte);
    }
}

static void print_sizes(void)
{
    printf("sizeof(char *)        = %zu\n", sizeof(char *));
    printf("sizeof(long)          = %zu\n", sizeof(long));
    printf("sizeof(int)           = %zu\n", sizeof(int));
    printf("sizeof(short)         = %zu\n", sizeof(short));
    printf("sizeof(mode_t)        = %zu\n", sizeof(mode_t));
    printf("sizeof(branchcount_t) = %zu\n", sizeof(branchcount_t));
    printf("sizeof(cvstime_t)     = %zu\n", sizeof(cvstime_t));
    printf("sizeof(time_t)        = %zu\n", sizeof(time_t));
    printf("sizeof(cvs_number)    = %zu\n", sizeof(cvs_number));
    printf("sizeof(node_t)        = %zu\n", sizeof(node_t));
    printf("sizeof(cvs_symbol)    = %zu\n", sizeof(cvs_symbol));
    printf("sizeof(cvs_branch)    = %zu\n", sizeof(cvs_branch));
    printf("sizeof(cvs_version)   = %zu\n", sizeof(cvs_version));
    printf("sizeof(cvs_patch)     = %zu\n", sizeof(cvs_patch));
    printf("sizeof(nodehash_t)    = %zu\n", sizeof(nodehash_t));
    printf("sizeof(editbuffer_t)  = %zu\n", sizeof(editbuffer_t));
    printf("sizeof(cvs_file)      = %zu\n", sizeof(cvs_file));
    printf("sizeof(rev_master)    = %zu\n", sizeof(rev_master));
    printf("sizeof(revdir)        = %zu\n", sizeof(revdir));
    printf("sizeof(cvs_commit)    = %zu\n", sizeof(cvs_commit));
    printf("sizeof(git_commit)    = %zu\n", sizeof(git_commit));
    printf("sizeof(rev_ref)       = %zu\n", sizeof(rev_ref));
    printf("sizeof(rev_list)      = %zu\n", sizeof(rev_list));
    printf("sizeof(cvs_commit_list) = %zu\n", sizeof(cvs_commit_list));
    printf("sizeof(rev_diff)      = %zu\n", sizeof(rev_diff));
    printf("sizeof(cvs_author)    = %zu\n", sizeof(cvs_author));
    printf("sizeof(chunk_t)       = %zu\n", sizeof(chunk_t));
    printf("sizeof(Tag)           = %zu\n", sizeof(tag_t));
}

struct checkpoint {
    const char *legend;
    struct timespec timespec;
    struct rusage rusage;
};

static struct checkpoint checkpoints[5];
static int ncheckpoints;

static void gather_stats(const char *legend)
/* gather resource usage statistics */
{
    if (ncheckpoints >= sizeof(checkpoints)/sizeof(struct checkpoint)) {
	announce("cran out of statistics slots %s", legend);
	return;
    }

    checkpoints[ncheckpoints].legend = legend;
    (void)clock_gettime(CLOCK_REALTIME, &checkpoints[ncheckpoints].timespec);
    (void)getrusage(RUSAGE_SELF, &checkpoints[ncheckpoints].rusage);
    ncheckpoints++;
}

int
main(int argc, char **argv)
{
    typedef enum _execution_mode {
	ExecuteExport, ExecuteGraph, ExecuteAuthors,
    } execution_mode;

    execution_mode  exec_mode = ExecuteExport;
    forest_t        forest;
    export_options_t export_options = {
	.branch_prefix = "refs/heads/",
	.id_token_expand =  EXPANDUNSPEC,
    };
    export_stats_t	export_stats;

#if defined(__GLIBC__)
    /* 
     * The default sbrk call grabs memory from the OS in 128kb chunks
     * for malloc. As we use up memory in prodigous amounts it is
     * better to grab it in bigger chunks, and make less calls to
     * sbrk. 16MB seems to be a decent value.
     */
    mallopt(M_TOP_PAD,16*1024*1024); /* grab memory in 16MB chunks */
#endif /* __GLIBC__ */
    clock_gettime(CLOCK_REALTIME, &export_options.start_time);
    memset(&export_stats, '\0', sizeof(export_stats_t));

    /* force times using mktime to be interpreted in UTC */
    setenv("TZ", "UTC", 1);

    LOGFILE = stderr;

    while (1) {
	static const struct option options[] = {
	    { "help",		    0, 0, 'h' },
	    { "version",	    0, 0, 'V' },
	    { "verbose",	    0, 0, 'v' },
	    { "quiet",		    0, 0, 'q' },
	    { "commit-time-window", 1, 0, 'w' },
	    { "log",                1, 0, 'l' },
	    { "authormap",          1, 0, 'A' },
	    { "authorlist",         1, 0, 'a' },
	    { "revision-map",       1, 0, 'R' },
	    { "reposurgeon",        0, 0, 'r' },
            { "graph",              0, 0, 'g' },
            { "expand",             2, 0, 'k' },
            { "remote",             1, 0, 'e' },
            { "strip",              1, 0, 's' },
            { "progress",           0, 0, 'p' },
            { "promiscuous",        0, 0, 'P' },
            { "incremental",        1, 0, 'i' },
            { "threads",	    0, 0, 't' },
            { "canonical",          0, 0, 'C' },
            { "fast",               0, 0, 'F' },
            { "embed-id",           0, 0, 'E' },
	    { "sizes",              0, 0, 'S' },	/* undocumented */
	};
	int c = getopt_long(argc, argv, "+hVw:l:grvqaA:R:Tk:e:s:pPi:t:CFSE", options, NULL);
	if (c < 0)
	    break;
	switch(c) {
	case 'h':
	    printf("Usage: cvs-fast-export [OPTIONS] [FILE]...\n"
		   "Parse RCS files and emit a fast-import stream.\n\n"
                   "Mandatory arguments to long options are mandatory for short options too.\n"
                   " -h --help                       This help\n"
		   " -g --graph                      Dump the commit graph\n"
		   " -k --expand                     Enable keyword expansion\n"
                   " -V --version                    Print version\n"
                   " -w --commit-time-window=WINDOW  Time window for commits(seconds)\n"
		   " -a --authorlist                 Report committer IDs from repository\n"
		   " -A --authormap                  Author map file\n"
		   " -R --revision-map               Revision map file\n"
		   " -r --reposurgeon                Issue cvs-revision properties\n"
		   " -T                              Force deteministic dates\n"
                   " -e --remote                     Relocate branches to refs/remotes/REMOTE\n"
                   " -s --strip                      Strip the given prefix instead of longest common prefix\n"
		   " -p --progress                   Enable load-status reporting\n"
		   " -P --promiscuous                Process files without ,v extension\n"
		   " -v --verbose                    Show verbose progress messages\n"
		   " -q --quiet                      Suppress normal warnings\n"
		   " -i --incremental TIME           Incremental dump beginning after specified RFC3339-format time.\n"
		   " -t --threads N                  Use threaded scheduler for CVS master analyses.\n"
		   " -E --embed-id                   Embed CVS revisions in the commit messages.\n"
		   "\n"
		   "Example: find | cvs-fast-export\n");
	    return 0;
	case 'g':
	    exec_mode = ExecuteGraph;
	    break;
	case 'P':
	    import_options.promiscuous = true;
	    break;
        case 'k':
	    export_options.id_token_expand = expand_override(optarg);
	    break;
	case 'v':
	    import_options.verbose++;
#ifdef YYDEBUG
	    extern int yydebug;
	    yydebug = 1;
#endif /* YYDEBUG */
	    break;
	case 'q':
	    nowarn = true;
	    break;
	case 'V':
	    printf("%s: version " VERSION "\n", argv[0]);
	    return 0;
	case 'w':
	    assert(optarg);
	    commit_time_window = atoi(optarg);
	    break;
	case 'l':
	    assert(optarg);
	    LOGFILE = fopen(optarg, "w");
	    break;
	case 'a':
	    exec_mode = ExecuteAuthors;
	    break;
	case 'A':
	    assert(optarg);
	    load_author_map(optarg);
	    break;
	case 'R':
	    assert(optarg);
	    export_options.revision_map = fopen(optarg, "w");
	    if (export_options.revision_map == NULL)
		fatal_error("cannot open %s for revision-map write", optarg);
	    break;
	case 'r':
	    export_options.reposurgeon = true;
	    break;
	case 'E':
	    export_options.embed_ids = true;
	    break;
	case 'T':
	    export_options.force_dates = true;
	    break;
	case 'e':
	    assert(optarg);
	    export_options.branch_prefix = (char*)xmalloc(strlen(optarg)+15, __func__);
	    sprintf(export_options.branch_prefix, "refs/remotes/%s/", optarg);
	    break;
	case 's':
	    assert(optarg);
	    import_options.striplen = strlen(optarg) + 1;
	    break;
	case 'p':
	    progress = true;
	    break;
	case 'i':
	    assert(optarg);
	    export_options.fromtime = convert_date(optarg);
	    break;
	case 't':
#ifdef THREADS
	    assert(optarg);
	    threads = atoi(optarg);
#else
	    announce("not built with thread support, -t option ignored.\n");
#endif
	    break;
	case 'C':
	    export_options.reportmode = canonical;
	    break;
	case 'F':
	    export_options.reportmode = fast;
	    break;
	case 'S':
	    print_sizes();
	    return 0;
	default: /* error message already emitted */
	    announce("try `%s --help' for more information.\n", argv[0]);
	    return 1;
	}
    }

    if (export_options.reposurgeon) {
	if (export_options.embed_ids)
	    fatal_error("The options --reposurgeon and --embed-id cannot be combined.\n");
    }

    argv[optind-1] = argv[0];
    argv += optind-1;
    argc -= optind-1;

#ifdef THREADS
#ifdef _SC_NPROCESSORS_ONLN
    if (threads == NO_MAX)
	threads = 2 * sysconf(_SC_NPROCESSORS_ONLN);
#endif /*  _SC_NPROCESSORS_ONLN */
#endif

    gather_stats("before parsing");

    /* build CVS structures by parsing masters; may read stdin */
    analyze_masters(argc, argv, &import_options, &forest);

    gather_stats("after parsing");

    /* commit set coalescence happens here */
    forest.git = merge_to_changesets(forest.cvs, 
				     forest.filecount,
				     import_options.verbose);

    gather_stats("after branch merge");

    /* report on the DAG */
    if (forest.git) {
	switch(exec_mode) {
	case ExecuteGraph:
	    dump_rev_graph(forest.git, NULL);
	    break;
	case ExecuteAuthors:
	    export_authors(&forest, &export_options);
	    break;
	case ExecuteExport:
	    export_commits(&forest, &export_options, &export_stats);
	    if (export_options.revision_map != NULL)
		fclose(export_options.revision_map);
	    break;
	}
    }

    gather_stats("total");

    if (progress)
    {
	float elapsed;
	struct checkpoint *chp;

	for (chp = checkpoints+1; chp < checkpoints + ncheckpoints; chp++)
	    fprintf(STATUS, "%20s:\t%.3f\t%ldKB\n",
		   chp->legend,
		   seconds_diff(&chp->timespec, &checkpoints[0].timespec),
		   chp->rusage.ru_maxrss - checkpoints[0].rusage.ru_maxrss);

	elapsed = seconds_diff(&checkpoints[ncheckpoints-1].timespec, 
			       &checkpoints[0].timespec);
	fprintf(STATUS, "%ld commits/%.3fM text, %u atoms at %d commits/sec.\n",
		export_stats.export_total_commits,
		export_stats.snapsize / 1000000.0,
		natoms,
		(int)(export_stats.export_total_commits / elapsed));
    }

    if (LOGFILE != stderr) {
	if (warncount > 0)
	    fprintf(STATUS, "cvs-fast-export: %u warning(s).\n", warncount);
	fclose(LOGFILE);
    }

    discard_atoms();
    discard_tags();
    revdir_free();
    free_author_map();
    return forest.errcount > 0;
}

/* end */
